### PR TITLE
fix auras with strata=TOOLTIP highjacking cursor's clicks

### PR DESCRIPTION
**Pull request by mrbuds**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings

### CODE REVIEW

File: WeakAuras/RegionTypes/RegionPrototype.lua [hunk 1]
```diff
@@ -473,6 +473,18 @@ function WeakAuras.regionPrototype.AddExpandFunction(data, region, cloneId, pare
     WeakAuras.Animate("display", data, "main", data.animation.main, region, false, nil, true, cloneId);
   end
 
+  function region:OptionsClosed()
```
Comment (emptyrivers) [inline WeakAuras/RegionTypes/RegionPrototype.lua]: Our leeches who want an Options closed event will be happy about this, I think 😅

**Comment by mrbuds** [comment]
"OnMouseDown" script is added on region:Expand() if Options are opened.
I could not remove the script on region:Collapse() because WeakAuras.IsOptionsOpen() already return false.
So removal of the script is done on WeakAurasOptions.frame "OnHide" script on all regions

File: WeakAuras/RegionTypes/RegionPrototype.lua (hunk 1 above)
Comment (mrbuds) [reply WeakAuras/RegionTypes/RegionPrototype.lua]: :D

File: WeakAuras/RegionTypes/RegionPrototype.lua [hunk 2]
```diff
@@ -473,6 +473,21 @@ function WeakAuras.regionPrototype.AddExpandFunction(data, region, cloneId, pare
...
+    region:EnableMouse(false)
+    region:SetScript("OnMouseDown", nil)
+  end
+
+  function region:ClickToPick()
+    region:EnableMouse(true)
+    region:SetScript("OnMouseDown", function()
+      WeakAuras.PickDisplay(region.id, nil, true)
+    end)
+    if region.GetFrameStrata and region:GetFrameStrata() == "TOOLTIP" then
```
Comment (InfusOnWoW) [inline WeakAuras/RegionTypes/RegionPrototype.lua]: Is this still needed?

File: WeakAuras/RegionTypes/RegionPrototype.lua [hunk 3]
```diff
@@ -473,6 +473,21 @@ function WeakAuras.regionPrototype.AddExpandFunction(data, region, cloneId, pare
     WeakAuras.Animate("display", data, "main", data.animation.main, region, false, nil, true, cloneId);
   end
 
+  function region:OptionsClosed()
+    region:EnableMouse(false)
+    region:SetScript("OnMouseDown", nil)
+  end
+
+  function region:ClickToPick()
```
Comment (InfusOnWoW) [inline WeakAuras/RegionTypes/RegionPrototype.lua]: Where's the call to ClickToPick for existing buttons? Is that not needed?

File: WeakAuras/RegionTypes/RegionPrototype.lua (hunk 2 above)
Comment (mrbuds) [reply WeakAuras/RegionTypes/RegionPrototype.lua]: Yes, without the strata change when i make a cursor aura with frameStrata = "TOOLTIP" visible from WeakAurasOptions i can't click anything anymore.

```
!nwvBZjUnq4)oPDAdMq6LJBM(bGdUl3qaMiFV1jJbf71G6jKCLK5L8b)BV7UM364CZ0egXJF8Us7(S7ksANeNiwNiYOpi0zdYGYAsMeLi2nnp3dbgNAnzk6nE0Sa9S26q4K28NB41o86T0kAISmSY6MwCHxRvoh62ZjI9xS3oyjAs8(cireGDHshcKMu07zwLb9R)044PpmE4imeFwdMShSzOn9hpCY7Xih8s0jzaODwAuRRZb0yFq6ceiJEpsopWhJhOeYJNh)OXAao(Kk2no7EKYJdgKw6dwuNcoPXRXtk(1PhvAste54kDyxv4SlDG3)Bvl4i5BhbFhbzGoi)2rW3)LQpvTO(FTnvQrlXvOAXFwTOD1Iwvl(89ZKoWeE3hGWW8CapJnGGm6QMoVd349SZO1dkDEsm9Cj8sRDaQCg08Qf)6HtS2XJpv9juUrbJW7zfrQlwjp0uGYupMC)zI(mbd)adzF3XWD1fCQldyyPhMFs9s6)tluyXjxzu(v)VlMOhLoDIyviu4FxRwBLlTxRST(4p0UOSbz5TOCj9y35PwLlpjeJWA2WfnN3BcGlxMcp1lld9)PVcYF0dJk)tZSBbhdFaYuYNQzP176qnkQLlbNVEYj(ebbjPi36wtQaHL6TY9(GRK1LsdSbm1Jv2ZsqIaZIdVzW5zuHrUgom05lFMnquMNR2LiMpONiEUiU3J0HsPNOql3JbHWxaA99zhCKcc(99pVjZCaVjIzdhporSemGtLkwz3ofJfp)DpUX8y5a11qPNMsFgdG65MpoC8SrFEmFiMZsalYzk)FZZp0wi166I0giU2mEYFYV3oQU4G1bJu)fus5BTAFl1oSEdLmTVo66i6rD(HRrgmCs8Whr3YOrxAKGVylxT8qD3HQgnmhK1D68DqJi264wm4XHdNWXyko5aZ)p3xws77y1vFD02)5L7V9VseBofxNgBq0kqTCfE13B6EDx8V30(T)rNOBU70Cbk3ARKVpwzw6SLf8T0L6G6OirLgQHWrghWXltOHnPAP33GflXPniDwn0GmtH9lPiZ(M2JT(niRdJon4ZRhYAWxSP4Ne51030GxsnWWRKtQxAeoi9wvwyvYK7W2GOOoVTBu3UTVH)PjSD89WgS(qDmmvY)o
```

File: WeakAuras/RegionTypes/RegionPrototype.lua (hunk 3 above)
Comment (mrbuds) [reply WeakAuras/RegionTypes/RegionPrototype.lua]: I'm not sure if i understand your question.
region:ClickToPick() is called by region:Expand() if WeakAuras.IsOptionsOpen()
When clicking on an aura button or the visibility toggle, ClickToPick() is called by region:Expand()

File: WeakAuras/RegionTypes/RegionPrototype.lua (hunk 2 above)
Comment (InfusOnWoW) [reply WeakAuras/RegionTypes/RegionPrototype.lua]: That aura is using custom code to attach to the mouse frame. Using the standard and supported way to attach it to the cursor just happens to work perfectly.

**Comment by InfusOnWoW** [comment]
I'm not too keen on overwritting frame strata from the options. It looks like the code to rest the frame starta is missing?

File: WeakAuras/RegionTypes/RegionPrototype.lua [hunk 4]
```diff
@@ -525,7 +540,9 @@ function WeakAuras.regionPrototype.AddExpandFunction(data, region, cloneId, pare
       end
       region.justCreated = nil;
       region:SetFrameLevel(WeakAuras.GetFrameLevelFor(region.id));
-
+      if WeakAuras.IsOptionsOpen() then
```
Comment (InfusOnWoW) [inline WeakAuras/RegionTypes/RegionPrototype.lua]: I would prefer if expand doesn't need to check if the options are open. Was the are any problem with the old code for setting the script?

**Comment by emptyrivers** [comment]
We have precedent for overriding regions properties while the options panel is open, see here:  https://github.com/WeakAuras/WeakAuras2/blob/60e9ede2a25ef724a2bdcdd688aeeebf5c43a0f4/WeakAurasOptions/WeakAurasOptions.lua#L1475-L1481

**Comment by mrbuds** [comment]
region:Expand() set the strata on each call already, there is no need to restore it
```
region:SetFrameLevel(WeakAuras.GetFrameLevelFor(region.id));
```

File: WeakAuras/RegionTypes/RegionPrototype.lua (hunk 4 above)
Comment (mrbuds) [reply WeakAuras/RegionTypes/RegionPrototype.lua]: If my memory doesn't fail me, the clicking was working if you clicked on the green eye icon to show regions, but not if you clicked the buttons, i'll move the code back to the DisplayButton properly

Comment (mrbuds) [reply WeakAuras/RegionTypes/RegionPrototype.lua]: yeah indeed, after reverting the change, any modification made on an aura which call WeakAuras.Add break ClickToPick

Comment (mrbuds) [reply WeakAuras/RegionTypes/RegionPrototype.lua]: fixed with https://github.com/WeakAuras/WeakAuras2/pull/1270/commits/e2d3897b6c892758eb05ecd33b963f20f5c63c29

**Comment by Stanzilla** [comment]
minor, winkyface

**Comment by mrbuds** [comment]
with these last changes moving an icon leave a ghost icon before it's droped
img

**Comment by mrbuds** [comment]
LFTM

**Pull request merged**